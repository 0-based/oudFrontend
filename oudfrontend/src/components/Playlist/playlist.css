.playlist{
    background-color: rgb(0,0,0);
    padding: 30px 32px;
    color: #b3b3b3;
    position: relative;
    min-height: 800px;
}

.playlistHeader{
    margin-bottom: 20px;
    padding-bottom: 20px;
    max-height: 340px;
}

.playlistHeaderBodyTop{
    padding-bottom: 10px;
    text-align:center;
    margin-bottom: 10px;
}

.whiteText{
    color: #fff;
}

.playlistAnchor:hover{
    color: #fff;
    text-decoration: underline;
}
.playlistAnchor{
    color:#b3b3b3;
    margin-left: auto;
    margin-right: auto;
}



.playButton {
    border-radius: 500px;
    font-family: Arial, Helvetica, sans-serif;
    font-weight: bolder;
    font-size: 12px;
    line-height: 18px;
    background-color: #1db954;
    color: #ffffff;
    width: 200px;
    margin: 10px auto;
    padding: 8px 10px;
    align-self: flex-start;
    transition: transform 0.2s; /* Animation */
    border-color: transparent;
    max-width: 114px;
    letter-spacing: 1.75px;
    
}
  
  .playButton:hover {
    transform: scale(1.06);
    background-color: #1db954;
  }
  .playButton:active {
    background-color: #1db954;
    border: #1db954;
  }
  .playButton:focus{
      outline: none;
  }

.playlistHeaderBodyBottom{
    
    text-align:center;
}
.playlistImage{   
    max-width: 320px;
    max-height: 320px;
    margin: 10px auto;    
   
    width:100%;
}



.playlistImageContainer{
    position: relative;
    text-align:center;
    max-height: 340px;

}

.playlist {
    
    font-family: "Helvetica Neue";
    -webkit-background-size: cover;
    -moz-background-size: cover;
    -o-background-size: cover;
    background-size: cover;
    background-repeat: no-repeat;
    background: rgb(125, 126, 125);
    background: -moz-linear-gradient(
      -45deg,
      rgba(125, 126, 125, 1) 0%,
      rgba(14, 14, 14, 1) 100%
    );
    background: -webkit-linear-gradient(
      -45deg,
      rgba(125, 126, 125, 1) 0%,
      rgba(14, 14, 14, 1) 100%
    );
    background: linear-gradient(
      135deg,
      rgba(125, 126, 125, 1) 0%,
      rgba(14, 14, 14, 1) 100%
    );
    filter: progid:DXImageTransform.Microsoft.gradient( startColorstr='#7d7e7d', endColorstr='#0e0e0e',GradientType=1 );
    background-attachment: fixed;
  }
.likeIcon{
    margin: 10px;
    padding: 0px;
    background-color: transparent;
    border: none;
}

.likeIcon:focus{
    outline: none;
}