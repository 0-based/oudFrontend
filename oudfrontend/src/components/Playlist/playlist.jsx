import React from 'react';
import './playlist.css';
import axios from 'axios';
import HeaderBodyBottom from '../commonComponents/headerBodyBottom'
import HeaderBodyTop from './components/headerBodyTop'
import SongList from '../commonComponents/songList'
import PropTypes from 'prop-types';



/**
 * @classdesc this is a component that renders playlist page
 * @author Ahmed Walid <ahmedwa1999@gmail.com>
 * @class
 * @param {string} id the Id of the playlist
 * @property {object} state carries the state of the component
 * @property {boolean} state.recived true if the data of the playlist is fetched correctly false otherwise
 * @property {boolean} state.liked true if the playlists is liked by the user (i.e the playlist is in the likedPlaylists table in the database)
 * @property {object} state.playlist carries all the information of the playlist
 * @property {Array.<track>} state.tracks array of all the songs in the playlist
 * @property {boolean} playing true when the playist is playing. Otherwise, it is false
 * @property {boolean} queued true when the playist is added to queue. Otherwise, it is false
 * @returns {
 *              <div>
 *               <div classname="playlistHeader">
 *                  <div classname="imageContainer">
 *                      <img/>
 *                  </div>
 * 
 *                  <div classname="playlistHeaderBody">
 *                      <HeaderBodyTop/>
 *                      <HeaderBodyBottom/>
 *                  </div>
 *                  <SongList/>
 *               </div>
 *              </div>
 *          
 *          }
 */
class Playlist extends React.Component{

    /** 
     * @constructor
     */ 
    constructor(props){
        super(props);
        this.state = {
            tracks : [],
            recieved:false,
            playlist:{},
            liked:false,
            playing:false,
            queued:false
        };
        const id = this.props.id;
    }
    /**
     * Called Whenever the user clicked on the PLAY button and it adds all the songs of the playlist to the queue by a post request
     * @func
     * @returns {void}
     */
    playButtonClicked(){
        //all the three requests should be put requests
        this.setState({playing:!this.state.playing})
        if(this.state.queued === false){
            this.setState({queued:true});
            const tracks = this.state.tracks
            const length = this.state.tracks.length
            axios.post('http://localhost:3000/queue/', {
                tracks : tracks,
                total : length
            })
            .then(function (response) {
                console.log(response);
            })
            .catch(function (error) {
                console.log(error);
            });
        }
        if(this.state.playing === false){
            axios.post('http://localhost:3000/player/pause/',)
            .then(function (response) {
                console.log(response);
            })
            .catch(function (error) {
                console.log(error);
            });
        }
        if(this.state.playing === true){
            axios.post('http://localhost:3000/player/play/',)
            .then(function (response) {
                console.log(response);
            })
            .catch(function (error) {
                console.log(error);
            });
        }

    }

    /**
     * Called Whenever the user clicked on the like button and it adds the playlist to the likedPlaylists 
     * if it is not already there otherwise it removes it from there by a delete request
     * @func
     * @returns {void}
     */
    likeButtonClicked(){
        const likedPlaylist = this.state.playlist
        if(this.state.liked === false){
            this.setState({liked:true})
            axios.post('http://localhost:3000/likedPlaylists/', likedPlaylist)
            .then(function (response) {
                console.log(response);
            })
            .catch(function (error) {
                console.log(error);
            });
        }
        else{
            this.setState({liked:false})

            axios.delete(`http://localhost:3000/likedPlaylists/${this.props.id.id}`)
            .then(function (response) {
                console.log(response);
            })
            .catch(function (error) {
                console.log(error);
            });
        }

    }
    /**
     * It fetches the data of the playlist from the database and checks if it exists in the likedPlaylists table
     * @func
     * @returns {void}
     */
    componentDidMount(){
        axios.get(`http://localhost:3000/playlist/${this.props.id.id}`)
        .then((response)=> {
            const playlist = response.data;
            this.setState({tracks:playlist.tracks});
            this.setState({recieved:true})
            this.setState({playlist:playlist})   
        })
        .catch((error)=> {
            console.log(error);
        });  
        
        axios.get(`http://localhost:3000/likedPlaylists/${this.props.id.id}`)
        .then((response)=> {
            console.log(response);
            this.setState({liked:true})
        })
        .catch((error)=> {
            console.log(error);
        });
    }
    
    

    render(){
        return(
            <div data-testid='playlist' className='playlist'>
                
                <div className='row'>
                    <div data-testid="playlistHeader" className='playlistHeader row col-xs-4 col-md-6 col-lg-4 col-xl-4'>
                        <div data-testid="playlistIamgeContainer" className='playlistImageContainer col col-lg-12 col-md-12 col-sm-4 col-xs-4'>
                            <img data-testid="playlistIamge" src={this.state.playlist.image} className='playlistImage' alt='playlist img'/>
                        </div>            
                        <div data-testid="playlistHeaderBody" className='playlistHeaderBody col col-lg-12 col-md-12 col-sm-8 col-xs-8'>
                            <HeaderBodyTop 
                                data-testid="HeaderBodyTop"
                                title={this.state.playlist.name}
                                owner={this.state.playlist.owner}
                              />
                            <HeaderBodyBottom 
                                data-testid = "HeaderBodyBottom" 
                                length = {this.state.tracks.length} 
                                playClicked = {this.playButtonClicked.bind(this)}
                                likeClicked = {this.likeButtonClicked.bind(this)}
                                liked = {this.state.liked}
                                playing = {this.state.playing}
                            />
                        </div>
                    </div>  
                    <SongList 
                        data-testid="songList"
                        recieved = {this.state.recieved}
                        tracks={this.state.tracks} 
                        className="col-xs-8 col-md-6 col-lg-8 col-xl-8"
                    />
                    
                </div>
            </div>
        );
    }
}
Playlist.propTypes={
    id : PropTypes.objectOf(PropTypes.string)
};
export default Playlist;