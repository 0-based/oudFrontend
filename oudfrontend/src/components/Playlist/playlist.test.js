import React from 'React'
import renderer from 'react-test-renderer';
import Playlist from './playlist.jsx';
import Enzyme, {shallow} from 'enzyme'
import EnzymeAdapter from 'enzyme-adapter-react-16'
import checkPropTypes from 'check-prop-types'

Enzyme.configure({adapter: new EnzymeAdapter()});


const setup = (props={}) =>{
    return shallow(<Playlist id = {props}/>);
}

const findByTestAttr=(wrapper, val)=>{
    return wrapper.find(`[data-testid="${val}"]`);
}

const trueProps={id:"1"}



/*
Rendering tests
*/

describe('Playlist Component', ()=>{
    describe('testing playlist renders Correctly', ()=>{
        let component;
        beforeEach (()=>{
            component = setup(trueProps);
        })
        it('renders playlist component', ()=>{
            const wrapper = findByTestAttr(component, "playlist");
            expect(wrapper.length).toBe(1);
        });
        it('renders playlist component', ()=>{
            const wrapper = findByTestAttr(component, "playlistHeader");
            expect(wrapper.length).toBe(1);
        });
        it('renders playlist component', ()=>{
            const wrapper = findByTestAttr(component, "playlistIamge");
            expect(wrapper.length).toBe(1);
        });
        it('renders playlist component', ()=>{
            const wrapper = findByTestAttr(component, "playlistIamgeContainer");
            expect(wrapper.length).toBe(1);
        });
        it('renders playlist component', ()=>{
            const wrapper = findByTestAttr(component, "playlistHeaderBody");
            expect(wrapper.length).toBe(1);
        });
        it('renders playlist component', ()=>{
            const wrapper = findByTestAttr(component, "HeaderBodyBottom");
            expect(wrapper.length).toBe(1);
        });
        it('renders playlist component', ()=>{
            const wrapper = findByTestAttr(component, "HeaderBodyTop");
            expect(wrapper.length).toBe(1);
        });
        it('renders playlist component', ()=>{
            const wrapper = findByTestAttr(component, "songList");
            expect(wrapper.length).toBe(1);
        });
    
    
    });
    
    describe('snapshot test for the playlist', ()=>{
        it('renders correctly', () => {
            const tree = renderer
              .create(<Playlist id={trueProps}/>)
              .toJSON();
            expect(tree).toMatchSnapshot();
        });
    });
    describe('checking propTypes', ()=>{
        const propsT = {id:{id :'1'}}
        const propsF ={id:{id :true}}
        it('should not throw a warning', ()=>{
            const result = checkPropTypes(Playlist.propTypes, propsT, 'prop', Playlist.name);
            expect(result).toBeUndefined();
        });
        it('should throw a warning', ()=>{
            const result = checkPropTypes(Playlist.propTypes, propsF, 'prop', Playlist.name);
            expect(result).toBeDefined();
        });
    });
});



  