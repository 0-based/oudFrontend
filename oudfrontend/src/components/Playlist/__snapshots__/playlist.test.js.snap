// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Playlist Component snapshot test for the playlist renders correctly 1`] = `
<div
  className="playlist"
  data-testid="playlist"
>
  <div
    className="row"
  >
    <div
      className="playlistHeader row col-xs-4 col-md-6 col-lg-4 col-xl-4"
      data-testid="playlistHeader"
      onClick={[Function]}
    >
      <div
        className="playlistImageContainer col col-lg-12 col-md-12 col-sm-4 col-xs-4"
        data-testid="playlistIamgeContainer"
      >
        <img
          alt="playlist img"
          className="playlistImage"
          data-testid="playlistIamge"
        />
      </div>
      <div
        className="playlistHeaderBody col col-lg-12 col-md-12 col-sm-8 col-xs-8"
        data-testid="playlistHeaderBody"
      >
        <div
          className="playlistHeaderBodyTop"
          data-testid="HeaderBodyTop"
        >
          <h2
            className="whiteText"
            data-testid="title"
          />
          <span
            className="whiteText"
            data-testid="credits"
          >
            Crerated By 
          </span>
          <a
            className="playlistAnchor"
            data-testid="owner"
            href="/user/undefined"
            onClick={[Function]}
          >
            
          </a>
        </div>
        <div
          className="playlistHeaderBodyBottom"
          data-testid="HeaderBodyBottom"
        >
          <button
            className="playButton"
            data-testid="playButton"
            onClick={[Function]}
            variant="outline-success"
          >
            PLAY
          </button>
          <button
            className="likeIcon"
            data-testid="likeIcon"
            onClick={[Function]}
          >
            like
             
          </button>
          <p>
            <span
              data-testid="releaseDate"
            >
              
               
              <h2
                data-testid="separatingDot"
                style={
                  Object {
                    "display": "inline",
                  }
                }
              >
                
              </h2>
               
            </span>
            <span
              data-testid="songsNumber"
            >
              0
               
            </span>
            <span
              data-testid="songsLiteral"
            >
              song
            </span>
          </p>
        </div>
      </div>
    </div>
    <div
      className="col-xs-8 col-md-6 col-lg-8 col-xl-8"
      data-testid="songsList"
    >
      <h1
        data-testid="loading"
      >
        LOADING ...
      </h1>
    </div>
  </div>
</div>
`;
